File: 0006-zigzag-conversion /0006-zigzag-conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        # If there's only one row or the string length is less than or equal to numRows, return the original string.
        if numRows == 1 or numRows >= len(s):
            return s

        idx, d = 0, 1  # Initialize the current row index and direction flag.
        rows = [[] for _ in range(numRows)]  # Create a list of lists to hold characters for each row.

        # Loop through each character in the input string.
        for char in s:
            rows[idx].append(char)  # Append the character to the current row.
            # Toggle direction when reaching the first or last row.
            if idx == 0:
                d = 1  # Move down if at the top row.
            elif idx == numRows - 1:
                d = -1  # Move up if at the bottom row.
            idx += d  # Update the row index based on the current direction.

        # Convert each list of characters in rows to a string.
        for i in range(numRows):
            rows[i] = ''.join(rows[i])

        # Join all row strings together into a single string and return it.
        return ''.join(rows)

File: 0006-zigzag-conversion /0006-zigzag-conversion.java
class Solution {
    public String convert(String s, int numRows) {
        // If there's only one row, return the original string as no zigzag is needed.
        if (numRows == 1 || numRows >= s.length()) {
            return s;
        }

        // Array of StringBuilder to hold strings for each row.
        StringBuilder[] rows = new StringBuilder[numRows];
        for (int i = 0; i < numRows; i++) {
            rows[i] = new StringBuilder();
        }

        int i = 0;  // Current row index
        int k = 1;  // Direction flag (1 for down, -1 for up)

        // Loop through each character in the input string
        for (char c : s.toCharArray()) {
            rows[i].append(c); // Append character to the current row
            // Toggle direction when the first or last row is reached
            if (i == 0) {
                k = 1;  // Start moving down
            } else if (i == numRows - 1) {
                k = -1; // Start moving up
            }
            i += k; // Move to the next row based on direction
        }

        // Combine all the rows into one final result
        StringBuilder result = new StringBuilder();
        for (StringBuilder row : rows) {
            result.append(row); // Append each row's content to the final result
        }

        return result.toString();
    }
}
